### PR TITLE
Site Creation Domain Suggestions: handle when there are no suggestions

### LINKED ISSUE #8429: Site Creation > Suggested Domains: handle no suggestions failure.
### Expected behavior
- Initial search: 
    - A message should be displayed.
    - The search bar should be displayed.
- Secondary search:
    - A message should be displayed.
    - The previous search results should be cleared.

### Actual behavior
- When fetching suggested domains, if it fails the view simply does nothing. There is no indication to the user what happened. 
- If it is the initial search, the search bar is not present for the user to search again.
initial_search

- If it is a secondary search, the previous results are still displayed.
secondary_search

### Steps to reproduce the behavior
Enter a site title that results in no suggestions, either on the Site Details page or in the Domain search bar. (Currently, Japanese text returns no suggestions.)

### PR DESCRIPTION
**Pull request by ScoutHarris**
Fixes #8429 

To test: 
- Enter a site title that results in no domain suggestions, either on the Site Details page or in the Domain search bar. 
- Verify the view below is displayed.

NOTE: Japanese text returns no suggestions.

Initial search, no suggestions from site title:
initial_search

No suggestions from a search via search bar:
secondary_search

### CODE REVIEW

**Comment by benhuberman** [comment]
@ScoutHarris I'd say there's room here to aim for something slightly warmer, for example:
> **We couldn't find any available address with that name -- let's try another one.**
> Enter a new name above and we'll look for an address that matches it.

A couple of other thoughts:
- Using "name" here feels somewhat confusing to me -- what the users are setting up is their URL, but that doesn't have to match their site title, and in many cases -- where domains are already taken -- it won't be possible to match these at all. I wonder if it might be clearer if we used a more neutral alternative, like "words" or "keywords":
> **We couldn't find any available address with the word you entered -- let's try again.**
> Enter a different word and we'll look for an address that matches it.
(note: if they're allowed to enter multiple words, we could tweak the above to "words," of course.)

- I'd also tweak the placeholder copy in the search box, since "Type to get more suggestions" doesn't really help guide the user, and "more" suggests there are already a few options, even though there are no suggestions at this state. Would something like this work, maybe?
> Enter a word or two you'd like to see in your site address.

**Comment by ScoutHarris** [comment]
@benhuberman - thank you! I've changed the body text to your suggestions.

Just one minor thing - the search placeholder text I believe has already been implemented on the Android side.

@folletto , @theck13 , @hypest  - how do you feel about changing the placeholder text per Ben's suggestion?

 > I'd also tweak the placeholder copy in the search box, since "Type to get more suggestions" doesn't really help guide the user, and "more" suggests there are already a few options, even though there are no suggestions at this state. Would something like this work, maybe?

> Enter a word or two you'd like to see in your site address.

**Comment by folletto** [comment]
I'd keep "Type to get more suggestions" because:

1. There should always be suggestions on screen, if not, we need to fix it. I've thought a very basic idea to avoid issues here: we always append any search result on TOP of the existing list, so at the very minimum the first screen of suggestions is always there. Not perfect, but if other solutions are too complicated for v1, let's go for it.
2. "Enter a word or two you'd like to see in your site address." is not functionally correct, because the system will also return exactly the requested string. 

If we can find another option that doesn't suggest a different functionality I'm up for it tho. 

Note: Calypso uses "Enter a name or keyword" (but doesn't provide a first screen of options already, so it's not exactly the same).

**Comment by folletto** [comment]
Sorry if I stress this aspect but: it's _key_ that this UI is never empty.

_(Still I appreciate the message screen if it does as safeguard!)._

**Comment by ScoutHarris** [comment]
@nheagy - since the endpoint has been fixed, no suggestions doesn't occur anymore. I know you could figure this out, but maybe so you don't have to hunt around, for testing you can simply put the `failure` logic in the `success` block for `service.getDomainSuggestions` in `SiteCreationDomainsTableViewController`.

**Comment by benhuberman** [comment]
@ScoutHarris If there are always going to be suggestions on this screen, the placeholder copy is a lot less confusing to me -- that was my main issue with it. That said, I do like the added specificity @folletto mentions in the Calypso UI, so I wonder if a fusion of the two might work here:

> Enter a name or keyword for more suggestions.

**Comment by ScoutHarris** [comment]
@benhuberman  - my apologies. When I did this PR this scenario happened frequently. The endpoint was fixed yesterday, so this view _should_ never appear. I should have updated the PR to reflect that.

I'll let @folletto comment on the placeholder text.

**Comment by folletto** [comment]
> Enter a name or keyword for more suggestions.

I like the structure, but seems too long. Might easily end up as: "Enter a name or keyword for mo...".
Consider also translations...

**Comment by benhuberman** [comment]
What would you say is the hard cap on character limit here, @folletto ? I could try to work around that. My proposed one above is 45 chars vs 28 chars in the original. This one is at 37 chars and I think is still an improvement over the original:

> Type a word or a name for more ideas.

This one is 32 chars:
> Need more ideas? Type a keyword.

**Comment by folletto** [comment]
Hard caps are hard as we don't know translations, and a text like "iiiiiii" is shorter than "mmmmm" — I know, tricky. :D 

What if we combine the two?

> Type a keyword for more ideas

**Comment by benhuberman** [comment]
> What if we combine the two?

Sure -- "Type a keyword for more ideas" works for me.

**Comment by folletto** [comment]
Wonderful! 
Thank you Ben for indulging me! :)

**Comment by ScoutHarris** [comment]
@benhuberman , @folletto  - thanks guys! Text updated, as well as screenshots in the PR description.

File: WordPress/Classes/ViewRelated/NUX/SiteCreationDomainsTableViewController.swift [hunk 1]
```diff
@@ -141,17 +152,35 @@ extension SiteCreationDomainsTableViewController {
...
 
+    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+
+        if indexPath.section == Sections.suggestions.rawValue && noSuggestions == true {
+            // Calculate the height of the no results cell from the bottom of
+            // the search field to the screen bottom, minus some padding.
+            let searchFieldRect = tableView.rect(forSection: Sections.searchField.rawValue)
+            let searchFieldBottom = searchFieldRect.origin.y + searchFieldRect.height
+            let screenBottom = UIScreen.main.bounds.height
+            return screenBottom - searchFieldBottom - 60
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/NUX/SiteCreationDomainsTableViewController.swift]: The `60` should be in an enum or struct/

File: WordPress/Classes/ViewRelated/NUX/SiteCreationDomainsTableViewController.swift [hunk 2]
```diff
@@ -207,6 +243,45 @@ extension SiteCreationDomainsTableViewController {
...
+
+        noResultsViewController.view.frame = cell.frame
+        cell.contentView.addSubview(noResultsViewController.view)
+
+        addChildViewController(noResultsViewController)
+        noResultsViewController.didMove(toParentViewController: self)
+    }
+
+    func removeNoResultsFromView() {
+        noSuggestions = false
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/NUX/SiteCreationDomainsTableViewController.swift]: `removeNoResultsFromView` should also reset the value of `tableView.separatorStyle`

**Comment by folletto** [comment]
> Also, the error isn't readable when the keyboard is up.

Good point, ach! I missed that.

What do y'all think? Should we just skip the image in this screen then and have the text up top?

File: WordPress/Classes/ViewRelated/NUX/SiteCreationDomainsTableViewController.swift [hunk 3]
```diff
@@ -71,14 +76,21 @@ class SiteCreationDomainsTableViewController: NUXTableViewController {
...
+            self?.noSuggestions = false
             addSuggestions(suggestions)
         }) { [weak self] (error) in
             DDLogError("Error getting Domain Suggestions: \(error.localizedDescription)")
             self?.isSearching = false
             SVProgressHUD.dismiss()
+            self?.tableView.separatorStyle = .none
+            self?.noSuggestions = true
+            // reload to display the no results view.
+            self?.tableView.reloadSections(IndexSet(integer: Sections.suggestions.rawValue), with: .automatic)
```
[the remark this answers is not in the document]
Comment (ScoutHarris) [reply WordPress/Classes/ViewRelated/NUX/SiteCreationDomainsTableViewController.swift]: > avoid adding noSuggestions and just test siteTitleSuggestions.count and searchSuggestions.count directly?

I do prefer `noSuggestions`. I think it's more obvious and concise than sprinkling the code with `if siteTitleSuggestions.count == 0 && searchSuggestions.count == 0`.

However, I could change the variable definition to:
```
private var noSuggestions: Bool {
        return siteTitleSuggestions.count == 0 && searchSuggestions.count == 0
    }
```

> This also helps handle the case where the suggestion endpoint comes back with success but 0 items. Was the old hammer search coming back as an error?

0 results does come back as an error, so there is no "extra" case to consider.

[the remark this answers is not in the document]
Comment (nheagy) [reply WordPress/Classes/ViewRelated/NUX/SiteCreationDomainsTableViewController.swift]: That sounds *great*. I'm on board 👍

[the remark this answers is not in the document]
Comment (ScoutHarris) [reply WordPress/Classes/ViewRelated/NUX/SiteCreationDomainsTableViewController.swift]: OK, well that went to 💩 fast. Now the NoResultsView is showing at odd times. I'll keep toying with it, but I don't think having `noSuggestions` calculated based on `siteTitleSuggestions.count` and `searchSuggestions.count` is going to work.

**Comment by ScoutHarris** [comment]
@nheagy - regarding:

> the error isn't readable when the keyboard is up. 

The keyboard is now dismissed when there are no results. I also took the liberty of dismissing the keyboard when there _are_ results, as I found it annoying that the keyboard was covering some of the results.

File: WordPress/Classes/ViewRelated/NUX/SiteCreationDomainsTableViewController.swift (hunk 2 above)
Comment (ScoutHarris) [reply WordPress/Classes/ViewRelated/NUX/SiteCreationDomainsTableViewController.swift]: It's set in the success & failure blocks of `service.getDomainSuggestions`. Adding `tableView.separatorStyle = .singleLine` in `removeNoResultsFromView` actually made no difference.

**Comment by ScoutHarris** [comment]
OK @nheagy it's ready for review once again.

**Comment by nheagy** [comment]
Let's do this!

:shipit:

**Pull request merged**